### ISSUE TITLE
Change Payment - Error when changing subscription payment method to a new 3DS CC with UPE disabled

### ISSUE THREAD

**Issue by veljkho** [Automattic/woocommerce-payments#2822]
### Describe the bug

In My Account > My Subscriptions > Change Payment page, when trying to add a new payment method which is a credit card with 3ds security, getting error: `Error: No such payment_intent: 'seti_ab'`. 

This occurs when the store has UPE **disabled**.

### To Reproduce

[pre-requisite: customer logged in and has 1 subscription bought (1 month subscription)]
1. Ensure store **doesn't** have UPE enabled ([using WCPay dev tools](https://github.com/Automattic/woocommerce-payments-dev-tools))
2. Go to `My Account` > `My Subscription`
3. Click on `Change Payment` button
4. Scroll down and check the checkbox to use a new payment method
5. Enter this card: `4000002760003184` 12/25 123
6. Click `Change Payment` button to save the changes
7. Observe issue

### Actual behavior

Throws warning and CC not added

### Screenshots

<img width="940" alt="image" src="https://user-images.githubusercontent.com/3147296/171344847-00de20d7-25ac-42d5-af05-e5d900af9066.png">

Page reloads and shows `Payment method updated` and spinner before the 3DS modal is rendered/validated!

<img width="1476" alt="Screen Shot 2022-07-28 at 3 10 23 PM" src="https://user-images.githubusercontent.com/4167300/181412237-028e8b5e-b2d1-4bc5-ae0c-abbc0f09691b.png">

<img width="1476" alt="Screen Shot 2022-07-28 at 3 10 25 PM" src="https://user-images.githubusercontent.com/4167300/181412255-df82ebc1-57a8-40c7-89ca-85c3a347820b.png">

After the 3DS modal is verified (click `Complete Auth`), then you get the error. It looks like our payment processing is firing too early.

<img width="1282" alt="Screen Shot 2022-07-28 at 3 10 04 PM" src="https://user-images.githubusercontent.com/4167300/181412364-b5558134-1bf2-4eec-b6cd-245a464ed98c.png">

### Expected behavior

No warning, credit card added successfully and present at the bottom.

### Desktop (please complete the following information):

* OS: MacOS Big Sur
* Browser Opera
* Version 78.0.4093.147

**Comment by james-allan** [comment]
duplicate of #1294

**Comment by htdat** [comment]
> duplicate of #1294

Thanks James! I am going to close #1294 as the current issue is more well-defined.

**Comment by haszari** [comment]
I think this is worth prioritising. If shoppers / subscribers can't [add 3DS (aka SCA)](https://stripe.com/docs/payments/3d-secure) credit cards as payment methods, we should either fix that or improve the error handling (if these cards are not supported for subscriptions).

**Comment by haszari** [comment]
I'm assuming this affects both WCPay Subscriptions **and** the integration with WooCommerce Subscriptions.

**Comment by Jinksi** [comment]
Testing on a fresh JN site after completing WC and WCPay onboarding in test mode:

✅ I can change a subscription payment method to a 3DS card successfully **once** after an initial card payment is charged.
❌ Subsequently changing the payment method to a new 3DS card via *My Account  → Subscriptions* will result in a processing error on checkout. See stripe error response below.
✅ Adding the new 3DS card via *My Account  → Payment methods* and then changing to it via *My Account  → Subscriptions* works as expected.
✅ Removing the previous 3DS card via *My Account  → Payment methods* and then changing payment to a 3DS card via *My Account  → Subscriptions* works as expected.
✅ I can make an initial purchase of a subscription product using a 3DS card.

**Test cards used**
- [SCA card](https://stripe.com/docs/development/quickstart#manual-tests) `4000002760003184` = modal displayed (same card as in the issue description)
- [3DS2 card](https://stripe.com/docs/payments/3d-secure#three-ds-cards) `4000000000003220`= modal displayed
- [3DS card](https://stripe.com/docs/payments/3d-secure#three-ds-cards) `4000000000003063` = no modal


<details>
<summary>Stripe error response</summary>

```json
{
	"error": {
		"code": "setup_intent_unexpected_state",
		"doc_url": "https://stripe.com/docs/error-codes/setup-intent-unexpected-state",
		"message": "You cannot update this SetupIntent because it has already succeeded.",
		"setup_intent": {
			"id": "seti_xxxxxxxxxxxxx",
			"object": "setup_intent",
			"cancellation_reason": null,
			"client_secret": "seti_xxxxxxxxxxxxxxxx",
			"created": 1653536655,
			"description": null,
			"last_setup_error": null,
			"livemode": false,
			"next_action": null,
			"payment_method": "pm_xxxxxxxxxxxxxxx",
			"payment_method_types": [
				"card"
			],
			"status": "succeeded",
			"usage": "off_session"
		},
		"type": "invalid_request_error"
	}
}
```

</details>

**Comment by Jinksi** [comment]
> I'm assuming this affects both WCPay Subscriptions and the integration with WooCommerce Subscriptions.

@haszari yes, following the above steps with WC Subscriptions produces the same results.

**Comment by Jinksi** [comment]
✅ After logging out of WP and back in, the payment method can be updated to a new 3DS card. **This is due to "setup intent" remaining from the previous payment request**:

https://github.com/Automattic/woocommerce-payments/blob/c0bb63a34f0014f223b875d7b6ae9394f614d959/client/checkout/classic/upe.js#L277-L278

If the following `if` statement is removed and a new "setup intent" is forced, then the change in payment method to a 3DS card succeeds on subsequent attempts.

https://github.com/Automattic/woocommerce-payments/blob/c0bb63a34f0014f223b875d7b6ae9394f614d959/client/checkout/classic/upe.js#L284-L290

**Comment by Jinksi** [comment]
I've moved the related bug with UPE to a new issue (#4308) since this issue is specific to non-UPE.

**Comment by haszari** [comment]
Tested this and I notice the form gets submitted, and Woo thinks a payment method has been added, before the 3DS modal is displayed. Eg.

- Shopper adds card details, submits form
- Page reloads, shows success message and error message (screenshots in description)
- 3DS modal is rendered here
- This is unlikely to work …

I compared with the regular (shortcode) checkout for WCPay, and pay-for-order for WC Stripe extension. In both cases, the 3DS modal is displayed before any page reloads. So maybe we need to rework the logic to ensure the 3DS modal is displayed before the form submission. 

The JS code handling this UI is in `checkout/classic/index.js`: https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/classic/index.js#L373

**Comment by haszari** [comment]
This could be the culprit:

https://github.com/Automattic/woocommerce-payments/blob/6081086475445df83dab2e3deda8f7555725938f/client/checkout/classic/index.js#L319-L322

- If we return `false` from the `submit` handler, it stops Woo from proceeding (i.e. the early page reload before 3DS modal).
- `handlePaymentMethodCreation` is the core of our submit handling.
- The first time it's called, it fires off an AJAX request to generate a payment method, and returns false to prevent submit.
- In the AJAX success callback, it sets `paymentMethodGenerated = true`. 
  - In the 3DS case this might not be the whole story.
- Then `handlePaymentMethodCreation` is called a second time, and this time it exits early (above code snippet) and allows checkout to proceed.
  - My hunch is that this needs to check something else and wait for the 3DS modal to proceed - either in the success callback, or in the subsequent call to `handlePaymentMethodCreation`.

**Comment by solstudioim** [comment]
Any update on this @haszari a merchant is affected by this, I think on 8894172-zd-a8c
The user is on WooPayments version 8.6.1

**Comment by haszari** [comment]
@solstudioim No updates from me or Helix team (if there were we would have commented!). I suggest reaching out to product or engineering folks currently focused on subscriptions, we switched focus a while ago.

**Comment by mattallan** [comment]
This issue might be similar to a bug we fixed in Stripe related to changing to a 3DS card after we implemented deferred intents: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3539

Deferred intents changed when payment/setup intents were created and confirmed. Previously on Stripe, the intent was created on page load and then confirmed at the time of form submission. With deferred intents, the intent is now created on form submission and later validated/confirmed during the `process_payment()` function which broke out compatibility with this feature (not sure if WooPayments does things the same).

In order to fix this issue, we had to return `false` on the [woocommerce_subscriptions_update_payment_via_pay_shortcode filter.](https://href.li/?https://github.com/Automattic/woocommerce-subscriptions-core/blob/6bfbf519df799dde8cd9f39e6c93e820e5170352/includes/class-wc-subscriptions-change-payment-gateway.php#L321-L322) which prevents WooCommerce Subscriptions from updating the payment method before the customer had authed the payment and the intent was confirmed.

**Comment by timur27** [comment]
This issue existed for a long time now, as it was there with the good old UPE too. The PR mentioned by Matt above is pretty helpful. On top of that, we still need to handle a couple of WooPayments-specific areas: 
1. Allow order meta to be updated for the change payment method use-case to ensure intent IDs alignment
2. Enhance 3DS intent confirmation AJAX to get the correct type of intent (our flow currently is not able to recognize it's a "change pm for subscription" use-case and attempts to fetch payment intents instead
3. Add the correct redirect as currently, assuming the above two points are addressed, I'm redirected to order received page

There might be more things in it but the above ones are the bigger ones for now.
